wai-extra/Network/Wai/Header.hs

```diff
@@ -14,5 +21,64 @@ contentLength hdrs = lookup H.hContentLength hdrs >>= readInt
 readInt :: S8.ByteString -> Maybe Integer
 readInt bs =
     case S8.readInteger bs of
-        Just (i, rest) | S8.null rest -> Just i
+        -- 'S8.all' is also 'True' for an empty string
+        Just (i, rest) | S8.all (== ' ') rest -> Just i
         _ -> Nothing
+
+replaceHeader :: H.HeaderName -> S.ByteString -> [H.Header] -> [H.Header]
+replaceHeader name val old =
+    (name, val) : filter ((/= name) . fst) old
+
+-- | Used to split a header value which is a comma separated list
+splitCommas :: S.ByteString -> [S.ByteString]
+splitCommas = map trimWS . S.split _comma
+
+-- Trim whitespace
+trimWS :: S.ByteString -> S.ByteString
+trimWS = dropWhileEnd (== _space) . S.dropWhile (== _space)
+
+-- | Dropping all 'Word8's from the end that satisfy the predicate.
+dropWhileEnd :: (Word8 -> Bool) -> S.ByteString -> S.ByteString
+#if MIN_VERSION_bytestring(0,10,12)
+dropWhileEnd = S.dropWhileEnd
+#else
+dropWhileEnd p = fst . S.spanEnd p
+#endif
+
+-- | Only to be used on header's values which support quality value syntax
+--
+-- A few things to keep in mind when using this function:
+-- * The resulting 'Int' will be anywhere from 1000 to 0 ("1" = 1000, "0.6" = 600, "0.025" = 25)
+-- * The absence of a Q value will result in 'Just 1000'
+-- * A bad parse of the Q value will result in a 'Nothing', e.g.
+--   * Q value has more than 3 digits behind the dot
+--   * Q value is missing
+--   * Q value is higher than 1
+--   * Q value is not a number
+parseQValueList :: S8.ByteString -> [(S8.ByteString, Maybe Int)]
+parseQValueList = fmap go . splitCommas
+  where
+    go = checkQ . S.break (== _semicolon)
+    checkQ :: (S.ByteString, S.ByteString) -> (S.ByteString, Maybe Int)
+    checkQ (val, "") = (val, Just 1000)
+    checkQ (val, bs) =
+        -- RFC 7231 says optional whitespace can be around the semicolon.
+        -- So drop any before it       ,           . and any behind it       $ and drop the semicolon
+        (dropWhileEnd (== _space) val, parseQval . S.dropWhile (== _space) $ S.drop 1 bs)
+      where
+        parseQval qVal = do
+            q <- S8.stripPrefix "q=" qVal
+            (i, rest) <- S8.uncons q
```

Comment:
It would be preferable to not use `S8.uncons` since this involves a conversion cost to a `Char`. Instead, using the codes for `'0`' and `'1'`, or https://www.stackage.org/package/word8, is ideal.